--- a/net/minecraft/world/item/crafting/CraftingRecipeCodecs.java
+++ b/net/minecraft/world/item/crafting/CraftingRecipeCodecs.java
@@ -15,10 +_,12 @@
       BuiltInRegistries.ITEM.byNameCodec(),
       p_301330_ -> p_301330_ == Items.AIR ? DataResult.error(() -> "Crafting result must not be minecraft:air") : DataResult.success(p_301330_)
    );
+   // TODO: add attachment support here
    public static final Codec<ItemStack> ITEMSTACK_OBJECT_CODEC = RecordCodecBuilder.create(
       p_301292_ -> p_301292_.group(
                ITEM_NONAIR_CODEC.fieldOf("item").forGetter(ItemStack::getItem),
-               ExtraCodecs.strictOptionalField(ExtraCodecs.POSITIVE_INT, "count", 1).forGetter(ItemStack::getCount)
+               ExtraCodecs.strictOptionalField(ExtraCodecs.POSITIVE_INT, "count", 1).forGetter(ItemStack::getCount),
+               ExtraCodecs.strictOptionalField(ExtraCodecs.withAlternative(net.minecraft.nbt.TagParser.AS_CODEC, net.minecraft.nbt.CompoundTag.CODEC), "nbt").forGetter(stack -> java.util.Optional.ofNullable(stack.getTag()))
             )
             .apply(p_301292_, ItemStack::new)
    );
